src/lib/Inference.hs

```diff
@@ -125,10 +125,105 @@ inferTopUDecl decl@(ULet _ (UPatAnn p ann) rhs) result = do
     applyDefaults
     return val
   return $ UDeclResultWorkRemaining block $ Abs decl result
-inferTopUDecl (UEffectDecl _ _ _) _ = throw NotImplementedErr "inferTopUDecl::UEffectDecl"
-inferTopUDecl (UHandlerDecl _ _ _ _ _ _) _ = throw NotImplementedErr "inferTopUDecl::UHandlerDecl"
+inferTopUDecl (UEffectDecl opTys effName opNames) result = do
+  opTys' <- forM opTys $ \opTy -> liftInfererM $ solveLocal $ inferEffOpType opTy
+  let effSourceName = uBinderSourceName effName
+  let effOpSourceNames = nestToList uBinderSourceName opNames
+  let effDef = EffectDef effSourceName (zip effOpSourceNames opTys')
+  effName' <- emitEffectDef effDef
+  opNames' <-
+    forM (enumerate effOpSourceNames) \(i, opName) -> do
+      emitEffectOpDef effName' i opName
+  let subst = effName @> effName' <.> opNames @@> opNames'
+  UDeclResultDone <$> applySubst subst result
+-- TODO(alex): handle retEff
```

Comment:
What is retEff?

Reply:
handlers can, themselves, add effects to the caller of the handler because the effect-op definitions might call effect operations. `retEff` is that effect row